src/measurement/vmsa.rs

```diff
@@ -0,0 +1,369 @@
+                ap_eip,
+                sev_features,
+                vcpu_type,
+                vmm_type,
+                cpu_num,
+            ))
+        } else {
+            None
+        };
+
+        VMSA {
+            bsp_save_area,
+            ap_save_area,
+        }
+    }
+
+    /// Generate a save area
+    fn build_save_area(
+        eip: u64,
+        sev_features: u64,
+        vcpu_type: CpuType,
+        vmm_type: VMMType,
+        cpu_num: Option<u64>,
+    ) -> SevEsSaveArea {
+        let mut area = SevEsSaveArea::default();
+
+        let (cs_flags, ss_flags, tr_flags, rdx) = match vmm_type {
+            VMMType::QEMU => (0x9b, 0x93, 0x8b, vcpu_type.sig() as u64),
+            VMMType::EC2 => {
+                if eip == 0xfffffff0 {
+                    (0x9a, 0x92, 0x83, 0)
+                } else {
+                    (0x9b, 0x92, 0x83, 0)
+                }
+            }
+            VMMType::KRUN => {
+                match cpu_num {
+                    Some(num) => {
+                        if num > 0 {
+                            area.rip = 0;
+                            area.rsp = 0;
+                            area.rbp = 0;
+                            area.rsi = 0;
+
+                            area.cs.selector = 0x9100;
+                            area.cs.base = 0x91000;
+                        } else {
+                            area.rsi = 0x7000;
+                            area.rbp = 0x8ff0;
+                            area.rsp = 0x8ff0;
+                        }
+                    }
+                    None => {
+                        area.rsi = 0x7000;
+                        area.rbp = 0x8ff0;
+                        area.rsp = 0x8ff0;
+                    }
+                };
+                (0x9a, 0x92, 0x83, 0)
+            }
+        };
+
+        area.es = VmcbSeg::new(0, 0x93, 0xffff, 0);
+        area.cs = VmcbSeg::new(0xf000, cs_flags, 0xffff, eip & 0xffff0000);
+        area.ss = VmcbSeg::new(0, ss_flags, 0xffff, 0);
+        area.ds = VmcbSeg::new(0, 0x93, 0xffff, 0);
+        area.fs = VmcbSeg::new(0, 0x93, 0xffff, 0);
+        area.gs = VmcbSeg::new(0, 0x93, 0xffff, 0);
+        area.gdtr = VmcbSeg::new(0, 0, 0xffff, 0);
+        area.idtr = VmcbSeg::new(0, 0, 0xffff, 0);
+        area.ldtr = VmcbSeg::new(0, 0x82, 0xffff, 0);
+        area.tr = VmcbSeg::new(0, tr_flags, 0xffff, 0);
+        area.efer = 0x1000;
+        area.cr4 = 0x40;
+        area.cr0 = 0x10;
+        area.dr7 = 0x400;
+        area.dr6 = 0xffff0ff0;
+        area.rflags = 0x2;
+        area.rip = eip & 0xffff;
+        area.g_pat = 0x7040600070406;
+        area.rdx = rdx;
+        area.sev_features = sev_features;
+        area.xcr0 = 0x1;
+
+        area
+    }
+
+    /// Return a vector containing the save area pages
+    pub fn pages(&self, vcpus: usize) -> Result<Vec<Vec<u8>>, MeasurementError> {
+        let bsp_page = bincode::serialize(&self.bsp_save_area)
```

Comment:
I believe we should be able to just use the `?` operator here instead of `map_err`

Reply:
Same comment as bincode::deserialize. The error is boxed and so mapping it allows us to access the acutal error.